heartbeat/scheduler/scheduler.go

```diff
@@ -18,118 +18,158 @@
 package scheduler
 
 import (
+	"context"
 	"errors"
-	"runtime/debug"
+	"fmt"
+	"math"
 	"sync"
 	"time"
 
+	"github.com/elastic/beats/heartbeat/scheduler/timerqueue"
+
+	"golang.org/x/sync/semaphore"
+
 	"github.com/elastic/beats/libbeat/common/atomic"
 	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/monitoring"
 )
 
 const (
 	statePreRunning int = iota + 1
 	stateRunning
-	stateDone
+	stateStopped
 )
 
+// InvalidTransitionError is returned from start/stop when making an invalid state transition, say from preRunning to stopped
+var InvalidTransitionError = fmt.Errorf("invalid state transition")
+
 type Scheduler struct {
-	limit uint
+	limit int64
 	state atomic.Int
 
 	location *time.Location
 
-	jobs   []*job
-	active uint // number of active entries
+	activeJobs         *monitoring.Uint // gauge showing number of active jobs
+	activeTasks        *monitoring.Uint // gauge showing number of active tasks
+	waitingTasks       *monitoring.Uint // number of tasks waiting to run, but constrained by scheduler limit
+	jobsPerSecond      *monitoring.Uint // rate of job processing computed over the past hour
+	jobsMissedDeadline *monitoring.Uint // counter for number of jobs that missed start deadline
```

Comment:
in order to reduce the 'bloat' of fields on the scheduler type how about:

```
type Scheduler struct {
   ...
   stats schedulerStats
   ...
}

type schedulerStats struct {
    activeJobs         *monitoring.Uint // gauge showing number of active jobs
	active uint // number of active entries
	activeTasks        *monitoring.Uint // gauge showing number of active tasks
	waitingTasks       *monitoring.Uint // number of tasks waiting to run, but constrained by scheduler limit
	jobsPerSecond      *monitoring.Uint // rate of job processing computed over the past hour
	jobsMissedDeadline *monitoring.Uint // counter for number of jobs that missed start deadline
}
```

Reply:
great call, will do